### PR TITLE
Improvements on measurement methodology of ZCounting

### PR DESCRIPTION
**Pull request by davidwalter2**
#### PR description:

Histograms are added to allow to correct measure the inner and outer muon track efficiencies. Further changes are done:
- Requirements on the inner and outer track pT, |eta|, and dR are now added in the global muon requirement; 
- The global muon efficiency is now measured using the merged standalone collection and using the outer track parameters; 
- The standalone muon efficiency is measured using only tracks that are not muon seeded and a dR matching to an outer track is performed (instead of taking the reference).

Only DQMOffline/Lumi will be affected by those changes

#### PR validation:

The code was validated using cmsDriver.py in CMSSW_12_4_12. It was also compared to a offline standalone implementation and identical results, as desired, were obtained.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

It should get merged in time for the 2023 data taking

### CODE REVIEW

**Comment by emanueleusai** [comment]
please test

**Comment by emanueleusai** [comment]
+1

**Comment by davidwalter2** [comment]
Hi @emanueleusai, is a backport needed for this to go into the release for 2023 data taking? 
Thanks,
David

File: DQMOffline/Lumi/plugins/ZCounting.cc [hunk 1]
```diff
@@ -463,17 +636,32 @@ void ZCounting::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup)
       if ((dilepMass < MassMin_) || (dilepMass > MassMax_))
         continue;
 
-      bool isTrackCentral = false;
-      if (fabs(eta2) < MUON_BOUND)
-        isTrackCentral = true;
+      // check if track is matched to standalone muon
+      bool isStandalone = false;
+      for (const reco::Track* itSta : hStandaloneProduct) {
+        if (reco::deltaR(itSta->eta(), itSta->phi(), eta2, phi2) < DRMAX_IO) {
```
Comment (perrotta) [inline DQMOffline/Lumi/plugins/ZCounting.cc:642]: use `deltaR2` instead

File: DQMOffline/Lumi/plugins/ZCounting.cc [hunk 2]
```diff
@@ -516,6 +700,28 @@ bool ZCounting::passMuonID(const reco::Muon& muon, const reco::Vertex* vtx) {
...
+bool ZCounting::passGlobalMuon(const reco::Muon& muon) {
+  // Global muon selection:
+  // - standard global muon criterium,
+  // - requirements on inner and outer track pT>15 and |eta|
+  // - requirements on deltaR(inner track, outer track)
+
+  return muon.isGlobalMuon() && muon.outerTrack()->numberOfValidHits() >= N_STA_HITS &&
+         muon.innerTrack()->pt() > MIN_PT_TRK && std::abs(muon.innerTrack()->eta()) < MAX_ETA_TRK &&
+         muon.outerTrack()->pt() > MIN_PT_STA && std::abs(muon.outerTrack()->eta()) < MAX_ETA_STA &&
+         reco::deltaR(
```
Comment (perrotta) [inline DQMOffline/Lumi/plugins/ZCounting.cc:713]: use `deltaR2` instead

File: DQMOffline/Lumi/plugins/ZCounting.cc [hunk 3]
```diff
@@ -306,17 +389,93 @@ void ZCounting::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup)
...
+  for (auto const& standAlone : *tracksStandAlone) {
+    auto const extraIdx = standAlone.extra().key();
+
+    const reco::Track* track = &standAlone;
+
+    // replicate logic in GlobalMuonProducer, take the updatedAtVtx track if it exists and has
+    // the same eta sign as the original, otherwise take the original
+    for (auto const& standAloneUpdatedAtVtx : *tracksStandAloneUpdatedAtVtx) {
+      if (standAloneUpdatedAtVtx.extra().key() == extraIdx) {
+        const int etaFlip1 = (standAloneUpdatedAtVtx.eta() * standAlone.eta()) < 0 ? -1 : 1;
```
Comment (perrotta) [inline DQMOffline/Lumi/plugins/ZCounting.cc:439]: Perhaps better
```suggestion
        const bool etaFlip1 = standAloneUpdatedAtVtx.eta() * standAlone.eta() >= 0;
```

File: DQMOffline/Lumi/plugins/ZCounting.cc [hunk 4]
```diff
@@ -399,27 +556,52 @@ void ZCounting::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup)
...
+          h_mass_ID_fail_BE->Fill(iEvent.luminosityBlock(), dilepMass);
         }
-      } else if (itMu2.isStandAloneMuon()) {
-        // category Sta: probe is a Standalone muon
+      }
+    }  // End of probe loop over muons
+
+    // Probe loop over standalone muons, for global muon efficiency calculation
+    for (std::vector<reco::Track>::size_type idx = 0; idx < hStandaloneProduct.size(); idx++) {
+      const reco::Track* itSta = hStandaloneProduct[idx];
```
Comment (perrotta) [inline DQMOffline/Lumi/plugins/ZCounting.cc:573]:
```suggestion
    for (const reco::Track* itSta : hStandaloneProduct) {
```

Comment (davidwalter2) [reply DQMOffline/Lumi/plugins/ZCounting.cc:573]: But I'm using idx in line 596

**Comment by perrotta** [comment]
please test

**Comment by emanueleusai** [comment]
> Hi @emanueleusai, is a backport needed for this to go into the release for 2023 data taking? Thanks, David

@davidwalter2 yes we need the 13_0 backport

**Comment by emanueleusai** [comment]
+1

**Comment by perrotta** [comment]
+1

**Pull request merged**